### ISSUE TITLE
SDK: Reduce size of SpanData struct

### ISSUE THREAD

**Issue by jmacd** [open-telemetry/opentelemetry-specification#77]
In the Java prototype, the SpanData struct has several fields that could be encoded in a more generic way.  Doing so benefits the implementations which can be smaller and simpler as a result, because they will have fewer SpanData fields to attend to.  We have two generic structures that can be taken advantage of.  Attributes will be useful when there is a one-to-one relationship with the span; events will be useful when there is a many-to-one relationship with the span.

Note this proposal rests on structured value encoding, discussed in issue https://github.com/open-telemetry/opentelemetry-specification/issues/76.

Specific simplifications that will reduce implementation code size:

1. *Links*. These are defined by a SpanContext and optional set of key-values.  These can be encoded using events to indicate the time that a link was established.  For example, a link from a parent to a child:

```
{
  "time": "...",          // when the child started
  "link": {
    "@type": "child_of",  // encodes a conventional child_of relationship
    "span_id": "1234",    // standard span identifiers
    "trace_id": "5678",
  }
}
```

It even becomes possible to treat the Event API itself as a way of recording ad-hoc trace linkage.  For example, to record an incident involving another span:

     my.Span.Event("stuck in queue: head of queue is", trace.IncidentalLink.Value(other.Span))

this might encode an event like:

```
{
  "time": "...",
  "message": "stuck in queue: head of queue is",
  "link": {
    "@type": "incident",
    "span_id": "other_span_id",
    "trace_id": "other_trace_id",
  }
}
```

2. *Kind*.  The span kind categorizes spans into distinct types.  These can be specified as semantic conventions for attributes, rather than explicitly coded fields.  For example,

```
{
  "span.kind": "client"  //  or "server", "internal", "consumer", "producer", etc.
}
```

This approach allows introducing new `Kind` values easily, which could be beneficial in cases where the predefined values do not exactly fit their situation.  I'm not sure if this is a case where vendors need the specification to be so prescriptive. Vendors, why do you need/want to know the span Kind?

3. *Status*.  The span status summaries the success or failure of a span.  These can be specified using a structured attribute value, for example:

```
  "span.status": {
    "@type": "google.com/proto/status",
    "code": "SERVER_ERROR",
    "message": "panic recovered: nil pointer exception"
  }
```

This approach could potentially allow more flexibility with regards to supporting other kinds of status object, which would make the OpenTelemetry specification less prescriptive for users. There is a balance to strike because, as vendors, we want to prescribe conventions that support useful product features. In this case, specifically, we are interested in knowing whether the span had an error or not; however it is not critical to prescribe that the error be encoded using a gRPC-style status object. In this sense, it may be better to specify that a "span.error" field, when set, indicates an error regardless of the object that is encoded there.

Then, users have the flexibility to encode span errors in any way they like.  For example, span errors can be simple strings:

```
{
  "span.error": "simple error message",
}
```

or they can be Java exceptions:

```
{
  "span.error": {
    "@type": "java.org/exception",
    "message": "SERVER_ERROR",
    "type": "java.lang.RuntimeException",
    "stacktrace": "...",
    "cause": {
      "@type": "java.org/exception",
      ...
    },
  }
}
```

**Comment by jmacd** [comment]
Remove `metrics` label. This is a trace data question, I think.

**Comment by songy23** [comment]
I think this issue will be N/A once https://github.com/open-telemetry/opentelemetry-specification/pull/215 is merged?

**Comment by jmacd** [comment]
Yes, I was just looking through issues w/ `metrics` label and found this miscategorized. Agree that it closes w/ #215.

**Comment by SergeyKanzhelev** [comment]
As @jmacd pointed above, this is not relevant now as SpanData was removed. Closing
